text/0000-fn-pin-put.md

```diff
@@ -70,13 +72,56 @@ For implementation-oriented RFCs (e.g. for compiler internals), this section sho
 # Reference-level explanation
 [reference-level-explanation]: #reference-level-explanation
 
-This is the technical portion of the RFC. Explain the design in sufficient detail that:
+We add a new trait, `FnPinMut`, with the signature above, to `std::ops`.
+It is subject to the same restrictions as the stable `Fn` traits:
+it cannot be named directly (use the `FnPinMut()` form),
+and it cannot be implemented by hand.
+(These restrictions may be lifted in the future.)
+
+An implementation of `FnPinMut` is provided for all closures where said implementation is sound.
+Practically, this means all currently stable `FnMut` closures, as they are `Unpin`, get the following implementation:
+
+```rust
+impl FnPinMut<Args> for $GeneratedClosureType {
+    extern "rust-call" fn call_pin(self: Pin<&mut Self>, args: Args) -> Self::Output {
+        (&mut *self).call_mut(args)
+    }
+}
+```
+
+Closures are now allowed to contain **yield expression**s of the form `yield $expr`.
+If a closure contains a yield expression, it _may not_ also contain a return expression.
+This includes the trailing return: a closure containing a `yield` must "return" `!`.
+This will most often be accomplished by a `panic!` or a `loop` at the tail end.
+In most ways (allowed positions, return type), a yield statement acts the same as a return expression.
 
-- Its interaction with other features is clear.
-- It is reasonably clear how the feature would be implemented.
-- Corner cases are dissected by example.
+We call a closure that uses `yield` instead of `return` a **yield closure**.
 
-The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+The key difference is that rather than execution restarting at the top of
+the closure on the next call, execution resumes after the yield statement.
+The compiler will generate a state machine to jump to the correct place upon each resumation.
+
+This is subject to the normal rules about what `Fn` traits are implemented!
+This means that a yield closure can implment all four traits: `Fn`, `FnMut`, `FnPinMut`, and `FnOnce`!
+For example, [`std::iter`'s `Counter` example](counter) can be reimplemented as simply
+
+```rust
+std::iter::from_fn(|| {
+    for count in 1..6 {
+        yield Some(count);
+    }
+    loop {
+        yield None;
+    }
+})
+```
+
+This closure will implement all four `Fn` traits.
```

Comment:
Is it correct that this closure can implement `Fn`?
A call modifies the state by iterating `1..6`. It needs a mutable reference.

Reply:
The only type of yield closure that can implement `Fn` is a yield closure with only 1 state.  For example, the yield closure `|x| loop { yield x + 1; }` could be `Fn` because it is exactly identical to `|x| x + 1`.